xbmc/platform/darwin/speech/SpeechRecognitionDarwin.mm

```diff
@@ -0,0 +1,325 @@
+                         [self onError:recognitionError logMessage:logMsg];
+                       }
+                       else
+                       {
+                         owner->OnRecognitionDone(listener);
+                       }
+
+                       self.recognitionRequest = nil;
+                       self.recognitionTask = nil;
+                       self.text = nil;
+                       [self.talkTimeoutTimer invalidate];
+                       self.talkTimeoutTimer = nil;
+                     }
+                   }];
+
+  [inputNode installTapOnBus:0
+                  bufferSize:4096
+                      format:[inputNode outputFormatForBus:0]
+                       block:^(AVAudioPCMBuffer* buffer, AVAudioTime* when) {
+                         [self.recognitionRequest appendAudioPCMBuffer:buffer];
+                       }];
+
+  [self.audioEngine prepare];
+  @try
+  {
+    NSError* outError;
+    [self.audioEngine startAndReturnError:&outError];
+
+    if (outError != nil)
+    {
+      [self onError:speech::RecognitionError::AUDIO
+          logMessage:[NSString stringWithFormat:
+                                   @"Audio engine couldn't start because of an error. code='%ld'",
+                                   outError.code]];
+      return;
+    }
+  }
+  @catch (NSException* exception)
+  {
+    [self onError:speech::RecognitionError::AUDIO
+        logMessage:[NSString stringWithFormat:
+                                 @"Audio engine couldn't start because of an excpetion. name='%@'",
+                                 exception.name]];
+    return;
+  }
+}
+
+- (void)speechRecognizer:(SFSpeechRecognizer*)speechRecognizer availabilityDidChange:(BOOL)available
+{
+  if (!available)
+  {
+    if (self.recognitionTask != nil)
+    {
+      [self.recognitionTask cancel];
+      self.recognitionTask = nil;
+    }
+
+    if (self.listener && self.owner)
+    {
+      [self onError:speech::RecognitionError::SERVICE_NOT_AVAILABLE
+          logMessage:@"Service currently not available. Try again later"];
+    }
+  }
+}
+
+- (void)onTalkTimeout:(NSTimer*)timer
+{
+  if (self.recognitionRequest != nil)
+    [self.recognitionRequest endAudio];
+}
+
+- (void)onError:(int)recognitionError logMessage:(NSString*)logMessage
+{
+  CLog::Log(LOGERROR, "Speech recognition error: {}", [logMessage UTF8String]);
+  self.listener->OnError(recognitionError);
+  self.owner->OnRecognitionDone(self.listener);
+}
+
+@end
+
+std::shared_ptr<speech::ISpeechRecognition> speech::ISpeechRecognition::CreateInstance()
+{
+  return std::make_shared<CSpeechRecognitionDarwin>();
+}
+
+void CSpeechRecognitionDarwin::StartSpeechRecognition(
+    const std::shared_ptr<speech::ISpeechRecognitionListener>& listener)
+{
+  // Speech: macOS 10.15+ iOS 10.0+. Currently not available on tvOS!
+  if (@available(macOS 10.15, iOS 10.0, *))
+  {
+    SpeechRecognitionImpl* recognizer = [[SpeechRecognitionImpl alloc] init];
```

Comment:
the created object isn't stored anywhere, it will die as soon as the internal timer is invalidated (the only strong reference). Is it intended?

Reply:
Ooops, this is not intended.

Reply:
How can I 'aquire' an instance to SpeechRecognitionImpl and release it (in CSpeechRecognitionDarwin dtor?